cluster/src/main/java/io/atomix/cluster/messaging/impl/NettyMessagingService.java

```diff
@@ -752,49 +768,146 @@ protected void initChannel(SocketChannel channel) throws Exception {
-          .addLast("decoder", new MessageDecoder())
-          .addLast("handler", dispatcher);
+      channel.pipeline().addLast("handshake", new ClientHandshakeHandlerAdapter(future));
+    }
+  }
+
+  /**
+   * Channel initializer for basic connections.
+   */
+  private class BasicServerChannelInitializer extends ChannelInitializer<SocketChannel> {
+    @Override
+    protected void initChannel(SocketChannel channel) throws Exception {
+      channel.pipeline().addLast("handshake", new ServerHandshakeHandlerAdapter());
+    }
+  }
+
+  /**
+   * Base class for handshake handlers.
+   */
+  private abstract class HandshakeHandlerAdapter extends ChannelInboundHandlerAdapter {
+    void writeProtocolVersion(ChannelHandlerContext context, ProtocolVersion version) {
+      ByteBuf buffer = context.alloc().buffer(6);
+      buffer.writeInt(preamble);
+      buffer.writeShort(version.version());
+      context.writeAndFlush(buffer);
+    }
+
+    Optional<ProtocolVersion> readProtocolVersion(ChannelHandlerContext context, ByteBuf buffer) {
+      int preamble = buffer.readInt();
+      if (preamble != NettyMessagingService.this.preamble) {
+        log.warn("Received invalid handshake, closing connection");
+        context.close();
+        return Optional.empty();
+      }
+
+      int version = buffer.readShort();
+      ProtocolVersion protocolVersion = ProtocolVersion.valueOf(version);
+      if (protocolVersion == null) {
+        context.close();
+      }
+      return Optional.ofNullable(protocolVersion);
+    }
+
+    void activateProtocolVersion(ChannelHandlerContext context, ProtocolVersion protocolVersion) {
+      MessagingProtocol protocol = protocolVersion.createProtocol(returnAddress);
+      context.pipeline().remove(this);
+      context.pipeline().addLast("encoder", protocol.newEncoder());
+      context.pipeline().addLast("decoder", protocol.newDecoder());
+      context.pipeline().addLast("handler", new InboundMessageDispatcher());
+    }
+  }
+
+  /**
+   * Client-side handshake handler adapter.
+   */
+  private class ClientHandshakeHandlerAdapter extends HandshakeHandlerAdapter {
+    private final CompletableFuture<Channel> future;
+
+    ClientHandshakeHandlerAdapter(CompletableFuture<Channel> future) {
+      this.future = future;
+    }
+
+    @Override
+    public void channelActive(ChannelHandlerContext context) throws Exception {
+      writeProtocolVersion(context, ProtocolVersion.latest());
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext context, Object message) throws Exception {
+      readProtocolVersion(context, (ByteBuf) message)
+          .ifPresent(version -> activateProtocolVersion(context, version));
+    }
+
+    @Override
+    void activateProtocolVersion(ChannelHandlerContext context, ProtocolVersion protocolVersion) {
+      super.activateProtocolVersion(context, protocolVersion);
+      future.complete(context.channel());
```

Comment:
The future needs to be completed exceptionally if an exception occurs in the pipeline.